### ISSUE TITLE
[BUG] Incorrect type inferred for pointer to function

### ISSUE THREAD

**Issue by dgrunwald** [cython/cython#4391]
**Describe the bug**
Inference determines the wrong type (one indirection too much) when taking the address of a function. This results in a cython compiler error.

**To Reproduce**
Code to reproduce the behaviour:
```cython
cdef extern from "header.h":
    void use(int)

    ctypedef void(* Func)(int)
    void with_callback(Func f)

def main():
    with_callback(&use)  # OK

    ptr = &use  # Cannot assign type 'void (*)(int)' to 'void (**)(int)'
    with_callback(ptr)  # Cannot assign type 'void (**)(int)' to 'Func'
```

**Expected behavior**
Type `void (*)(int)` should be inferred for `ptr`, allowing the program to compile successfully.

**Actual behavior**
```
Error compiling Cython file:
------------------------------------------------------------
...
    void with_callback(Func f)

def main():
    with_callback(&use)

    ptr = &use
         ^
------------------------------------------------------------

test.pyx:10:10: Cannot assign type 'void (*)(int)' to 'void (**)(int)'

Error compiling Cython file:
------------------------------------------------------------
...

def main():
    with_callback(&use)

    ptr = &use
    with_callback(ptr)
                 ^
------------------------------------------------------------

test.pyx:11:18: Cannot assign type 'void (**)(int)' to 'Func'
```

**Environment:**
 - OS: Windows
 - Python version: 3.9.1
 - Cython version: 0.29.24

**Comment by scoder** [comment]
Thanks for the report, that seems wrong.

The code for that is in
https://github.com/cython/cython/blob/97c05e7a936e1a8929ee8c3e2f6629cc061cb66b/Cython/Compiler/ExprNodes.py#L10512-L10513

Can't say right now how exactly this can be fixed since functions and their pointers are a bit special in C. So this needs a bit more investigation. We already infer the function pointer from the function name in the base class (`UnopNode.infer_type()`), and then creating a pointer from that leads to the duplication.

I added a test file for now, where we can put a test for this.
https://github.com/cython/cython/commit/97c05e7a936e1a8929ee8c3e2f6629cc061cb66b

Note that things like
```cython
f = &func
p = &f
p[0][0]
```
also need to keep working, so just inferring the function itself instead of the pointer isn't correct either.

Could you try to debug this a bit more?

**Comment by dgrunwald** [comment]
So what's happening in C is that `func` is something called a "function designator". It has type `void (int)`, i.e. it's not a pointer at all.
Taking the address of a function designator results in a function pointer.

However, C also has a little known feature called "function-to-function-pointer decay". It's the lesser known brother of array-to-pointer decay.
Basically, if you do anything with a function designator *except* taking its address, the C compiler will implicitly take the address of the function.

So `with_callback(use);` and `with_callback(&use);` are the same thing in C.
Even `with_callback(*use);` works and does the exact same thing! (The function designator `use` can't be dereferenced, so it implicitly turns into a function pointer which can. The result of the dereference is another function designator `*&use`, which then also implicitly decays to `&*&use`).
Basically, in C you can have an almost arbitary mixture of `&` and `*` in front of a function name. If there's two `&` in a row, you're taking the address of a function pointer rvalue, and that's a compiler error. But any other combination of `&` and `*` is valid and will (after decay) result in a single-level pointer to the function.

https://godbolt.org/z/qq6P16cP7

Cython already supports all of:
```
    with_callback(use)   # OK
    with_callback(&use)  # OK
    with_callback((&use)[0])  # OK
    with_callback(&(&use)[0])  # OK
 ```
So function designators seem to be supported by the cython compiler -- just not when inferring the type of a local variable.
